src/diggler/command.d

```diff
@@ -422,16 +424,129 @@ abstract class CommandSet(T) : ICommandSet
+		template isCommand(string symbol){
+			static if(__traits(getProtection, __traits(getMember, T, symbol)) != "public")
+				enum isCommand = false;
+			else
+				enum isCommand =
+					symbol != "__ctor" &&
+					symbol != "__dtor" &&
+					__traits(compiles, __traits(getMember, T, symbol)) && // ahem...
+					isSomeFunction!(__traits(getMember, T, symbol)) &&
+					!__traits(isStaticFunction, mixin("T." ~ symbol)) &&
+					!hasAttribute!(__traits(getMember, T, symbol), ignore);
+		}
+
+		//Filters a sorted sequence by prefix, e.g. prefixFilter!("a", "a_b", "a_b_c", "d_e_f") yields AliasSeq!("a_b", "a_b_c")
+		template prefixFilter(string prefix, sequence...)
+		{
+			template prefixFilterImpl(size_t pos, seq...)
+			{
+				static if(pos >= seq.length || !seq[pos].startsWith(prefix))
+					alias prefixFilterImpl = seq[0..pos];
+				else
+					alias prefixFilterImpl = prefixFilterImpl!(pos+1, seq);
+			}
+
+			alias prefixFilter = prefixFilterImpl!(0, sequence);
+		}
+
+		//Basic template to get the first part of a subcommand, e.g. rootCmd("a_b_c") is "a"
+		enum rootCmd(string name) = name.splitter(cmdSep).front;
+		//Count how "deep" a subcommand is, e.g. depth("a_b_c") is 3
+		enum depth(string name) = name.splitter(cmdSep).walkLength;
+
+		//Sort all of the valid commands in T
+		enum sortingFunc(string left, string right) = left < right;
+		alias sortedMembers = staticSort!(sortingFunc, Filter!(isCommand, __traits(derivedMembers, T)));
+
+		foreach(index, memberName; sortedMembers)
+		{
+			//Captures a subcommand
+			static if(!memberName.startsWith(cmdSep) && //Starting with cmdSep will cause problems
+				  (memberName.canFind(cmdSep) || //A subcommand either contains cmdSep, or is next to another command with the same rootCmd
+					((index + 1) < sortedMembers.length &&
+					 memberName.splitter(cmdSep).front == sortedMembers[index+1].splitter(cmdSep).front)))
+			{
+				//Ensure we havent already built a handler for this subcommand yet
+				static if(!(index && rootCmd!(sortedMembers[index-1]) == rootCmd!(memberName)))
+				{
+					Command createHandler(string prefix, Members...)()
+					{
+						//A prefix ending with the command separator will produce an unusable command, fail if we find it
+						static assert(!prefix.endsWith(cmdSep), "Command " ~ prefix ~ " cannot end with the command separator");
+
+						Command handler;
+						auto dg = (string s = null)
+						{
+							foreach(ref c; handler.subCommands)
+							{
+								if(s.startsWith(c.name))
+								{
+									if(!c.adminOnly || context.isAdmin(cast(string)context.user.nickName))
+										c.handler(!s ? s : s[c.name.length .. $]);
+									return;
+								}
+							}
+							static if(__traits(hasMember, T, prefix) && isCommand!prefix)
+								handler.eponymousCommand.handler(s);
+						};
+						handler = Command.create!((string s = null){})(dg);
+						handler.name = prefix.split(cmdSep)[$-1]; //can't use splitter(...).back
```

Comment:
This is `prefix.splitter(cmdSep).tail(1).front`; `tail` isn't in Phobos until 2.071 but it's just a loop:
```d
foreach(part; prefix.splitter(cmdSep))
    handler.name = part;
```